internal/authz/authz.go

```diff
@@ -0,0 +1,237 @@
+func (a *Authorizer) Authorize(ctx context.Context, clientID, storeID, apiMethod string) error {
+	relation, err := a.getRelation(apiMethod)
+	if err != nil {
+		return err
+	}
+
+	authorized, err := a.individualAuthorize(ctx, clientID, relation, a.getStore(storeID), &openfgav1.ContextualTupleKeys{})
+	if err != nil {
+		return err
+	}
+
+	if !authorized {
+		return status.Error(codes.Code(ErrorResponse.GetCode()), ErrorResponse.GetMessage())
+	}
+
+	return nil
+}
+
+// AuthorizeCreateStore checks if the user has access to create a store.
+func (a *Authorizer) AuthorizeCreateStore(ctx context.Context, clientID string) error {
+	relation, err := a.getRelation(CreateStore)
+	if err != nil {
+		return err
+	}
+	authorized, err := a.individualAuthorize(ctx, clientID, relation, a.getSystem(), &openfgav1.ContextualTupleKeys{})
+	if err != nil {
+		return err
+	}
+
+	if !authorized {
+		return status.Error(codes.Code(ErrorResponse.GetCode()), ErrorResponse.GetMessage())
+	}
+
+	return nil
+}
+
+// AuthorizeListStores checks if the user has access to list stores.
+func (a *Authorizer) AuthorizeListStores(ctx context.Context, clientID string) error {
+	relation, err := a.getRelation(ListStores)
+	if err != nil {
+		return err
+	}
+
+	authorized, err := a.individualAuthorize(ctx, clientID, relation, a.getSystem(), &openfgav1.ContextualTupleKeys{})
+	if err != nil {
+		return err
+	}
+	if !authorized {
+		return status.Error(codes.Code(ErrorResponse.GetCode()), ErrorResponse.GetMessage())
+	}
+
+	return nil
+}
+
+// ListAuthorizedStores returns the list of stores that the user has access to.
+func (a *Authorizer) ListAuthorizedStores(ctx context.Context, clientID string) ([]string, error) {
+	err := a.AuthorizeListStores(ctx, clientID)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &openfgav1.ListObjectsRequest{
+		StoreId:              a.config.StoreID,
+		AuthorizationModelId: a.config.ModelID,
+		User:                 a.getApplication(clientID),
+		Relation:             CanCallGetStore,
+		Type:                 StoreType,
+	}
+
+	// Disable authz check for the list objects request.
+	ctx = authcontext.ContextWithSkipAuthzCheck(ctx, true)
+	resp, err := a.server.ListObjects(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	authcontext.ContextWithSkipAuthzCheck(ctx, false)
+
+	return resp.GetObjects(), nil
+}
+
+func (a *Authorizer) individualAuthorize(ctx context.Context, clientID, relation, object string, contextualTuples *openfgav1.ContextualTupleKeys) (bool, error) {
+	req := &openfgav1.CheckRequest{
+		StoreId:              a.config.StoreID,
+		AuthorizationModelId: a.config.ModelID,
+		TupleKey: &openfgav1.CheckRequestTupleKey{
+			User:     a.getApplication(clientID),
+			Relation: relation,
+			Object:   object,
+		},
+		ContextualTuples: contextualTuples,
+	}
+
+	// Disable authz check for the check request.
+	ctx = authcontext.ContextWithSkipAuthzCheck(ctx, true)
+	resp, err := a.server.Check(ctx, req)
```

Comment:
The issue with calling Check on the server and not creating a command object for this is that you won't be able to configure what runs. We will be restricted to the global Check API settings: 

- throttling and caching may be turned on
- maxConcurrentReadsForCheck may be some low number like 30

I feel like the authorizer should not use any of these settings and should use as many resources as it needs.

@Vic-Dev no action item for you, just wanted to raise this to @openfga/backend

Reply:
> Caching will make the authorizer eventually consistent (is this desirable?)

Definitely yes - if anything we have higher tolerance for eventual consistency and caching here, if we can configure it separate from Check I would make it something like 1min

This is being called on every request so the more we can use the cache, the better

Reply:
Ok yes that makes sense. What about all the other settings? E.g. do we want to apply throttling to the authorizer?